File: api/services/income_frequency.py
"""Single source of truth for recurring-income / bill cadence normalization.

A recurring income (or bill) stores a **per-payment** amount; this maps a
cadence to the number of those payments in a month so a per-payment amount
can be normalized to a monthly figure (multiply) and a monthly figure can be
split into a per-payment amount (divide). The two operations are exact
inverses, so a salary entered as monthly and divided here round-trips back to
the same monthly income in `_monthly_income`.

CR note: "quincenal" is **semi-monthly** (paid on the 15th and month-end → 2
payments/month, 24/year), NOT bi-weekly (26/year). The factor is 2, so a
quincena is exactly half the monthly salary.
"""
from __future__ import annotations

from decimal import Decimal

# Payments per month by cadence. Used in BOTH directions:
#   monthly_amount = per_payment * factor   (normalization, _monthly_income)
#   per_payment    = monthly_amount / factor (salary entry division)
PAYMENTS_PER_MONTH: dict[str, Decimal] = {
    "weekly": Decimal("52") / Decimal("12"),
    "biweekly": Decimal("2"),  # CR quincenal = semi-monthly (2/month)
    "monthly": Decimal("1"),
    "annual": Decimal("1") / Decimal("12"),
}


def per_payment_from_monthly(monthly: Decimal, frequency: str) -> Decimal:
    """Split a monthly amount into the per-payment amount for `frequency`.

    Unknown cadence → returned unchanged (treated as monthly). Quantized to
    cents so the stored row matches what the user is shown.
    """
    factor = PAYMENTS_PER_MONTH.get(frequency)
    if factor is None or factor == 0:
        return monthly
    return (Decimal(monthly) / factor).quantize(Decimal("0.01"))

File: mobile/src/lib/incomeFrequency.ts
/**
 * Mirror of api/services/income_frequency.py — payments-per-month per cadence.
 *
 * A recurring income stores a PER-PAYMENT amount; the backend normalizes to a
 * monthly figure by multiplying by this factor. A salary is entered as a
 * MONTHLY figure, so the form divides by the factor before sending so the
 * round-trip preserves the monthly salary (a quincena = monthly / 2).
 *
 * CR: "quincenal" is semi-monthly (2 payments/month), NOT bi-weekly (26/yr).
 * Keep in sync with the Python constant.
 */
import type { IncomeFrequency } from "../api/incomes";

export const PAYMENTS_PER_MONTH: Record<IncomeFrequency, number> = {
  weekly: 52 / 12,
  biweekly: 2, // CR quincenal = semi-monthly
  monthly: 1,
  annual: 1 / 12,
};

/** Split a monthly amount into the per-payment amount for `frequency`. */
export function perPaymentFromMonthly(
  monthly: number,
  frequency: IncomeFrequency,
): number {
  const factor = PAYMENTS_PER_MONTH[frequency] ?? 1;
  return monthly / factor;
}

/** Reconstruct the monthly amount from a stored per-payment amount. */
export function monthlyFromPerPayment(
  perPayment: number,
  frequency: IncomeFrequency,
): number {
  const factor = PAYMENTS_PER_MONTH[frequency] ?? 1;
  return perPayment * factor;
}
